File: crates/rsvelte_capi/examples/java/Smoke.java
/*
 * rsvelte Java FFM smoke test.
 *
 * Requires JDK 22+ (FFM API finalized in java.lang.foreign).
 *
 * Run from the repository root:
 *   cargo build -p rsvelte_capi --release
 *   java --enable-native-access=ALL-UNNAMED \
 *        crates/rsvelte_capi/examples/java/Smoke.java
 *
 * This is a single-file source-launch program — `java Smoke.java`
 * compiles + runs in one shot (JEP 330) on modern JDKs.
 */

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemoryLayout;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SymbolLookup;
import java.lang.foreign.ValueLayout;
import java.lang.invoke.MethodHandle;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Locale;

public class Smoke {

    // RsvelteBuf { uint8_t* data; size_t len; size_t cap; }
    static final MemoryLayout BUF_LAYOUT = MemoryLayout.structLayout(
        ValueLayout.ADDRESS.withName("data"),
        ValueLayout.JAVA_LONG.withName("len"),
        ValueLayout.JAVA_LONG.withName("cap")
    );

    public static void main(String[] args) throws Throwable {
        Path repoRoot = Paths.get("").toAbsolutePath();
        Path dylib = repoRoot
            .resolve("target")
            .resolve("release")
            .resolve(dylibName());
        if (!dylib.toFile().exists()) {
            System.err.println("FAIL: dylib not found at " + dylib
                + " — run `cargo build -p rsvelte_capi --release` first");
            System.exit(2);
        }

        Linker linker = Linker.nativeLinker();
        SymbolLookup lib = SymbolLookup.libraryLookup(dylib, Arena.ofAuto());

        MethodHandle rsvelte_version = linker.downcallHandle(
            lib.find("rsvelte_version").orElseThrow(),
            FunctionDescriptor.of(ValueLayout.ADDRESS)
        );

        // void rsvelte_compile_into(const u8 *src, size_t src_len,
        //                           const u8 *opts, size_t opts_len,
        //                           RsvelteBuf *out)
        MethodHandle rsvelte_compile_into = linker.downcallHandle(
            lib.find("rsvelte_compile_into").orElseThrow(),
            FunctionDescriptor.ofVoid(
                ValueLayout.ADDRESS, ValueLayout.JAVA_LONG,
                ValueLayout.ADDRESS, ValueLayout.JAVA_LONG,
                ValueLayout.ADDRESS
            )
        );
        MethodHandle rsvelte_compile_module_into = linker.downcallHandle(
            lib.find("rsvelte_compile_module_into").orElseThrow(),
            FunctionDescriptor.ofVoid(
                ValueLayout.ADDRESS, ValueLayout.JAVA_LONG,
                ValueLayout.ADDRESS, ValueLayout.JAVA_LONG,
                ValueLayout.ADDRESS
            )
        );
        MethodHandle rsvelte_free_raw = linker.downcallHandle(
            lib.find("rsvelte_free_raw").orElseThrow(),
            FunctionDescriptor.ofVoid(
                ValueLayout.ADDRESS, ValueLayout.JAVA_LONG, ValueLayout.JAVA_LONG
            )
        );

        MemorySegment versionPtr = (MemorySegment) rsvelte_version.invoke();
        // The returned pointer has zero-length info; reinterpret with a
        // generous bound so we can read NUL-terminated bytes.
        String version = versionPtr.reinterpret(64)
            .getString(0, StandardCharsets.UTF_8);
        System.out.println("rsvelte version: " + version);

        int failures = 0;
        failures += run(
            rsvelte_compile_into, rsvelte_free_raw,
            "basic component",
            "<h1>Hello from Java, {name}!</h1>",
            null,
            true);
        failures += run(
            rsvelte_compile_into, rsvelte_free_raw,
            "runes + dev",
            "<script>let { count = 0 } = $props();</script>\n<button onclick={() => count++}>{count}</button>",
            "{\"filename\":\"App.svelte\",\"dev\":true,\"runes\":true}",
            true);
        failures += run(
            rsvelte_compile_into, rsvelte_free_raw,
            "ssr generate",
            "<p>server-rendered from java</p>",
            "{\"generate\":\"server\",\"filename\":\"Ssr.svelte\"}",
            true);
        failures += run(
            rsvelte_compile_module_into, rsvelte_free_raw,
            "module: $state rune",
            "export const counter = $state(0);",
            "{\"filename\":\"counter.svelte.js\"}",
            true);
        failures += run(
            rsvelte_compile_into, rsvelte_free_raw,
            "malformed options",
            "<h1>x</h1>",
            "{not json",
            false);

        if (failures == 0) {
            System.out.println("\nPASS — 0 failure(s)");
        } else {
            System.out.println("\nFAIL — " + failures + " failure(s)");
            System.exit(1);
        }
    }

    private static int run(
        MethodHandle compileFn,
        MethodHandle freeFn,
        String label,
        String source,
        String options,
        boolean expectOk
    ) throws Throwable {
        System.out.println("\n=== " + label + " ===");
        try (Arena arena = Arena.ofConfined()) {
            byte[] srcBytes = source.getBytes(StandardCharsets.UTF_8);
            MemorySegment srcSeg = arena.allocate(Math.max(srcBytes.length, 1));
            MemorySegment.copy(srcBytes, 0, srcSeg, ValueLayout.JAVA_BYTE, 0, srcBytes.length);

            MemorySegment optSeg;
            long optLen;
            if (options == null) {
                optSeg = MemorySegment.NULL;
                optLen = 0;
            } else {
                byte[] optBytes = options.getBytes(StandardCharsets.UTF_8);
                optSeg = arena.allocate(Math.max(optBytes.length, 1));
                MemorySegment.copy(optBytes, 0, optSeg, ValueLayout.JAVA_BYTE, 0, optBytes.length);
                optLen = optBytes.length;
            }

            MemorySegment outBuf = arena.allocate(BUF_LAYOUT);
            compileFn.invoke(
                srcBytes.length == 0 ? MemorySegment.NULL : srcSeg,
                (long) srcBytes.length,
                optSeg,
                optLen,
                outBuf
            );

            MemorySegment dataPtr = outBuf.get(ValueLayout.ADDRESS, 0);
            long len = outBuf.get(ValueLayout.JAVA_LONG, ValueLayout.ADDRESS.byteSize());
            long cap = outBuf.get(ValueLayout.JAVA_LONG, ValueLayout.ADDRESS.byteSize() + 8);

            if (dataPtr.equals(MemorySegment.NULL) || len == 0) {
                System.err.println("FAIL: " + label + " returned empty buffer");
                return 1;
            }
            MemorySegment view = dataPtr.reinterpret(len);
            String body = new String(view.toArray(ValueLayout.JAVA_BYTE), StandardCharsets.UTF_8);
            String preview = body.length() > 300 ? body.substring(0, 300) + "..." : body;
            System.out.println(preview);

            freeFn.invoke(dataPtr, len, cap);

            boolean isOk = body.startsWith("{\"ok\":true");
            boolean isFail = body.startsWith("{\"ok\":false");
            if (expectOk && !isOk) {
                System.err.println("FAIL: " + label + " expected ok=true");
                return 1;
            }
            if (!expectOk && !isFail) {
                System.err.println("FAIL: " + label + " expected ok=false");
                return 1;
            }
            return 0;
        }
    }

    private static String dylibName() {
        String osName = System.getProperty("os.name", "").toLowerCase(Locale.ROOT);
        if (osName.contains("mac") || osName.contains("darwin")) return "librsvelte_capi.dylib";
        if (osName.contains("win"))                                return "rsvelte_capi.dll";
        return "librsvelte_capi.so";
    }
}

File: crates/rsvelte_capi/examples/ruby/smoke.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# rsvelte Ruby smoke test via stdlib Fiddle (no gems).
#
# Run from the repository root:
#   cargo build -p rsvelte_capi --release
#   ruby crates/rsvelte_capi/examples/ruby/smoke.rb
#
# Fiddle's C parser doesn't understand returning struct-by-value, so
# we use the rsvelte_compile_into / rsvelte_compile_module_into
# out-parameter variants instead.

require "fiddle"
require "fiddle/import"
require "fiddle/struct"
require "json"
require "rbconfig"

ROOT       = File.expand_path("../../../..", __dir__)
TARGET_DIR = File.join(ROOT, "target", "release")
DYLIB_NAME = case RbConfig::CONFIG["host_os"]
             when /darwin/ then "librsvelte_capi.dylib"
             when /mswin|mingw|cygwin/ then "rsvelte_capi.dll"
             else "librsvelte_capi.so"
             end
DYLIB_PATH = File.join(TARGET_DIR, DYLIB_NAME)

unless File.exist?(DYLIB_PATH)
  warn "FAIL: dylib not found at #{DYLIB_PATH} — run `cargo build -p rsvelte_capi --release` first"
  exit 2
end

module Rsvelte
  extend Fiddle::Importer
  dlload DYLIB_PATH

  RsvelteBuf = struct ["void* data", "size_t len", "size_t cap"]

  extern "const char *rsvelte_version()"
  extern "void rsvelte_compile_into(const void *, size_t, const void *, size_t, void *)"
  extern "void rsvelte_compile_module_into(const void *, size_t, const void *, size_t, void *)"
  # Decomposed-args variant of rsvelte_free — the struct-by-value
  # variant uses a calling convention that Fiddle (Ruby 2.6) can't
  # construct correctly on AArch64.
  extern "void rsvelte_free_raw(void *, size_t, size_t)"
end

def call_into(fn, source, options_json)
  src = (source || "").b
  opt = (options_json || "").b

  src_ptr = src.empty? ? Fiddle::NULL : Fiddle::Pointer[src]
  opt_ptr = opt.empty? ? Fiddle::NULL : Fiddle::Pointer[opt]
  src_len = src.bytesize
  opt_len = opt.bytesize

  out = Rsvelte::RsvelteBuf.malloc
  begin
    fn.call(src_ptr, src_len, opt_ptr, opt_len, out)
    raise "rsvelte returned empty buffer" if out.len.zero? || out.data.to_i.zero?
    Fiddle::Pointer.new(out.data.to_i)[0, out.len]
  ensure
    Rsvelte.rsvelte_free_raw(out.data, out.len, out.cap) unless out.data.to_i.zero?
  end
end

def run_case(label:, source:, options:, expect_ok:, fn:)
  puts "\n=== #{label} ==="
  options_json = options.nil? ? "" : JSON.generate(options)
  body = call_into(fn, source, options_json)
  env  = JSON.parse(body)
  short = body.length > 300 ? body[0, 300] + "..." : body
  puts short
  if env["ok"] != expect_ok
    warn "FAIL: #{label} — expected ok=#{expect_ok}, got #{env['ok'].inspect}"
    return false
  end
  true
end

COMPILE        = Rsvelte.method(:rsvelte_compile_into)
COMPILE_MODULE = Rsvelte.method(:rsvelte_compile_module_into)

puts "rsvelte version: #{Rsvelte.rsvelte_version.to_s}"

failures = 0
failures += 1 unless run_case(
  label: "basic component",
  source: "<h1>Hello from Ruby, {name}!</h1>",
  options: nil,
  expect_ok: true,
  fn: COMPILE,
)
failures += 1 unless run_case(
  label: "runes + dev",
  source: "<script>let { count = 0 } = $props();</script>\n<button onclick={() => count++}>{count}</button>",
  options: { "filename" => "App.svelte", "dev" => true, "runes" => true },
  expect_ok: true,
  fn: COMPILE,
)
failures += 1 unless run_case(
  label: "ssr generate",
  source: "<p>server-rendered from ruby</p>",
  options: { "generate" => "server", "filename" => "Ssr.svelte" },
  expect_ok: true,
  fn: COMPILE,
)
failures += 1 unless run_case(
  label: "module: $state rune",
  source: "export const counter = $state(0);",
  options: { "filename" => "counter.svelte.js" },
  expect_ok: true,
  fn: COMPILE_MODULE,
)
# Error path — pass raw malformed JSON, bypassing the JSON.generate helper.
puts "\n=== malformed options ==="
body = call_into(COMPILE, "<h1>x</h1>", "{not json")
env  = JSON.parse(body)
puts body
unless env["ok"] == false
  warn "FAIL: malformed options — expected ok=false"
  failures += 1
end

if failures.zero?
  puts "\nPASS — 0 failure(s)"
  exit 0
end
puts "\nFAIL — #{failures} failure(s)"
exit 1
